File: apps/web/src/lib/kyc/service.ts
import { eq } from "drizzle-orm";
import { db } from "@/lib/db";
import {
  organizationVerificationApplications,
  organizations,
} from "@/lib/db/schema";
import {
  createPersonaInquiry,
  getPersonaInquiry,
  mapPersonaStatusToProviderStatus,
  mapPersonaStatusToVerificationStatus,
  personaInquiryMustBeReplaced,
  personaInquiryNeedsSessionToken,
  resumePersonaInquiry,
  type PersonaInquiryStatus,
} from "@/lib/kyc/persona";
import { normalizeCountryCode } from "@/lib/kyc/country";
import { VERIFICATION_VALIDITY_DAYS } from "@/constants/kyc";
import { getMembershipForUser } from "@/lib/organizations/members";

export class KycServiceError extends Error {
  constructor(
    message: string,
    readonly code: "forbidden" | "not_found" | "conflict" | "invalid"
  ) {
    super(message);
    this.name = "KycServiceError";
  }
}

function addVerificationValidityDays(from = new Date()) {
  const expires = new Date(from);
  expires.setDate(expires.getDate() + VERIFICATION_VALIDITY_DAYS);
  return expires;
}

function extractProfileFromPersonaFields(
  fields: Record<string, { type: string; value: unknown }> | null | undefined,
) {
  if (!fields) {
    return { displayName: null, country: null };
  }

  const read = (key: string) => {
    const value = fields[key]?.value;
    return typeof value === "string" && value.trim() ? value.trim() : null;
  };

  const firstName = read("name-first");
  const lastName = read("name-last");
  const fullName =
    read("name-full") ||
    read("full-name") ||
    [firstName, lastName].filter(Boolean).join(" ").trim() ||
    null;
  const country =
    read("address-country-code") ||
    read("selected-country-code") ||
    read("country-code") ||
    read("country") ||
    null;

  return { displayName: fullName, country };
}

export async function getVerificationApplicationForOrganization(organizationId: string) {
  return db.query.organizationVerificationApplications.findFirst({
    where: eq(organizationVerificationApplications.organizationId, organizationId),
  });
}

type VerificationApplication = NonNullable<
  Awaited<ReturnType<typeof getVerificationApplicationForOrganization>>
>;

async function replacePersonaInquiry(input: {
  organizationId: string;
  application: VerificationApplication;
  note?: string;
}) {
  const personaInquiry = await createPersonaInquiry({
    referenceId: input.organizationId,
    note:
      input.note ??
      `${input.application.accountType === "business" ? "Business" : "Personal"} account | New Persona inquiry`,
  });

  const now = new Date();
  const [application] = await db
    .update(organizationVerificationApplications)
    .set({
      providerInquiryId: personaInquiry.inquiryId,
      providerStatus: mapPersonaStatusToProviderStatus(personaInquiry.status),
      updatedAt: now,
    })
    .where(eq(organizationVerificationApplications.id, input.application.id))
    .returning();

  await db
    .update(organizations)
    .set({
      verificationStatus: "pending",
      verifiedAt: null,
      verificationExpiresAt: null,
      updatedAt: now,
    })
    .where(eq(organizations.id, input.organizationId));

  return application;
}

async function syncPersonaInquiryState(
  organizationId: string,
  application: VerificationApplication,
) {
  const inquiry = await getPersonaInquiry(application.providerInquiryId!);
  const profile = extractProfileFromPersonaFields(inquiry.fields);

  if (profile.displayName || profile.country) {
    await db
      .update(organizationVerificationApplications)
      .set({
        ...(profile.displayName ? { displayName: profile.displayName } : {}),
        ...(profile.country
          ? { country: normalizeCountryCode(profile.country) }
          : {}),
        updatedAt: new Date(),
      })
      .where(eq(organizationVerificationApplications.id, application.id));
  }

  await applyPersonaInquiryStatus({
    organizationId,
    inquiryId: inquiry.inquiryId,
    status: inquiry.status,
  });

  return inquiry;
}

async function resumePersonaInquiryWithFallback(input: {
  organizationId: string;
  application: VerificationApplication;
  inquiryId: string;
  inquiryStatus: PersonaInquiryStatus;
}) {
  try {
    const sessionToken = await resumePersonaInquiry(input.inquiryId);
    return {
      inquiryId: input.inquiryId,
      sessionToken,
    };
  } catch (error) {
    if (
      input.inquiryStatus === "created" ||
      personaInquiryMustBeReplaced(input.inquiryStatus)
    ) {
      throw error;
    }

    const application = await replacePersonaInquiry({
      organizationId: input.organizationId,
      application: input.application,
      note: "Replacement inquiry after Persona session resume failed",
    });

    const sessionToken = await resumePersonaInquiry(application.providerInquiryId!);

    return {
      inquiryId: application.providerInquiryId!,
      sessionToken,
    };
  }
}

export async function startVerification(input: {
  organizationId: string;
  userId: string;
  accountType?: "personal" | "business";
}) {
  const membership = await getMembershipForUser(input.organizationId, input.userId);

  if (!membership || membership.role !== "owner") {
    throw new KycServiceError("Only the business owner can start verification", "forbidden");
  }

  const organization = await db.query.organizations.findFirst({
    where: eq(organizations.id, input.organizationId),
  });

  if (!organization) {
    throw new KycServiceError("Business not found", "not_found");
  }

  const existing = await getVerificationApplicationForOrganization(input.organizationId);

  if (existing?.providerStatus === "approved") {
    throw new KycServiceError("This business is already verified", "conflict");
  }

  if (existing?.providerInquiryId) {
    const hasActiveInquiry =
      existing.providerStatus === "created" ||
      existing.providerStatus === "pending" ||
      existing.providerStatus === "needs_review";

    if (hasActiveInquiry) {
      return existing;
    }
  }

  const accountType = input.accountType ?? existing?.accountType ?? "personal";
  const displayName = organization.name;
  const countryCode = "XX";

  const personaInquiry = await createPersonaInquiry({
    referenceId: input.organizationId,
    note: `${accountType === "business" ? "Business" : "Personal"} account | Pending Persona details`,
  });

  const now = new Date();
  const applicationValues = {
    accountType,
    displayName,
    registrationNumber: null,
    country: countryCode,
    businessDescription: null,
    provider: "persona",
    providerInquiryId: personaInquiry.inquiryId,
    providerStatus: mapPersonaStatusToProviderStatus(personaInquiry.status),
    updatedAt: now,
  };

  let application;

  if (existing) {
    [application] = await db
      .update(organizationVerificationApplications)
      .set(applicationValues)
      .where(eq(organizationVerificationApplications.id, existing.id))
      .returning();
  } else {
    [application] = await db
      .insert(organizationVerificationApplications)
      .values({
        organizationId: input.organizationId,
        ...applicationValues,
      })
      .returning();
  }

  await db
    .update(organizations)
    .set({
      verificationStatus: "pending",
      verifiedAt: null,
      verificationExpiresAt: null,
      updatedAt: now,
    })
    .where(eq(organizations.id, input.organizationId));

  return application;
}

export async function syncVerificationFromPersona(organizationId: string) {
  const application = await getVerificationApplicationForOrganization(organizationId);

  if (!application?.providerInquiryId) {
    throw new KycServiceError("Verification has not been started yet", "not_found");
  }

  const inquiry = await getPersonaInquiry(application.providerInquiryId);
  const profile = extractProfileFromPersonaFields(inquiry.fields);

  if (profile.displayName || profile.country) {
    await db
      .update(organizationVerificationApplications)
      .set({
        ...(profile.displayName ? { displayName: profile.displayName } : {}),
        ...(profile.country
          ? { country: normalizeCountryCode(profile.country) }
          : {}),
        updatedAt: new Date(),
      })
      .where(eq(organizationVerificationApplications.id, application.id));
  }

  await applyPersonaInquiryStatus({
    organizationId,
    inquiryId: inquiry.inquiryId,
    status: inquiry.status,
  });

  return getVerificationSummary(organizationId);
}

export async function applyPersonaInquiryStatus(input: {
  organizationId: string;
  inquiryId: string;
  status: Parameters<typeof mapPersonaStatusToProviderStatus>[0];
}) {
  const application = await getVerificationApplicationForOrganization(input.organizationId);

  if (!application || application.providerInquiryId !== input.inquiryId) {
    return null;
  }

  const providerStatus = mapPersonaStatusToProviderStatus(input.status);
  const verificationStatus = mapPersonaStatusToVerificationStatus(input.status);
  const now = new Date();

  await db
    .update(organizationVerificationApplications)
    .set({
      providerStatus,
      updatedAt: now,
    })
    .where(eq(organizationVerificationApplications.id, application.id));

  if (verificationStatus === "verified") {
    const verifiedAt = now;
    const verificationExpiresAt = addVerificationValidityDays(verifiedAt);

    await db
      .update(organizations)
      .set({
        verificationStatus: "verified",
        verifiedAt,
        verificationExpiresAt,
        updatedAt: verifiedAt,
      })
      .where(eq(organizations.id, input.organizationId));

    return { verificationStatus: "verified" as const };
  }

  if (verificationStatus === "rejected") {
    await db
      .update(organizations)
      .set({
        verificationStatus: "rejected",
        verifiedAt: null,
        verificationExpiresAt: null,
        updatedAt: now,
      })
      .where(eq(organizations.id, input.organizationId));

    return { verificationStatus: "rejected" as const };
  }

  await db
    .update(organizations)
    .set({
      verificationStatus: "pending",
      updatedAt: now,
    })
    .where(eq(organizations.id, input.organizationId));

  return { verificationStatus: "pending" as const };
}

export async function getVerificationSession(organizationId: string, userId: string) {
  const membership = await getMembershipForUser(organizationId, userId);

  if (!membership) {
    throw new KycServiceError("Forbidden", "forbidden");
  }

  let application = await getVerificationApplicationForOrganization(organizationId);

  if (!application?.providerInquiryId) {
    throw new KycServiceError("Start verification before opening the Persona flow", "invalid");
  }

  let inquiry = await syncPersonaInquiryState(organizationId, application);

  application = await getVerificationApplicationForOrganization(organizationId);

  if (!application?.providerInquiryId) {
    throw new KycServiceError("Verification has not been started yet", "not_found");
  }

  if (personaInquiryMustBeReplaced(inquiry.status)) {
    application = await replacePersonaInquiry({
      organizationId,
      application,
      note: "Replacement inquiry after Persona declined or failed",
    });
    inquiry = await getPersonaInquiry(application.providerInquiryId);
  }

  let inquiryId = application.providerInquiryId;
  let sessionToken: string | null = null;

  if (personaInquiryNeedsSessionToken(inquiry.status)) {
    const resumed = await resumePersonaInquiryWithFallback({
      organizationId,
      application,
      inquiryId,
      inquiryStatus: inquiry.status,
    });
    inquiryId = resumed.inquiryId;
    sessionToken = resumed.sessionToken;
  }

  return {
    inquiryId,
    sessionToken,
  };
}

export async function getVerificationSummary(organizationId: string) {
  const organization = await db.query.organizations.findFirst({
    where: eq(organizations.id, organizationId),
  });

  if (!organization) {
    throw new KycServiceError("Business not found", "not_found");
  }

  const application = await getVerificationApplicationForOrganization(organizationId);

  const now = new Date();
  const isExpired =
    organization.verificationExpiresAt != null &&
    organization.verificationExpiresAt <= now;

  return {
    organization,
    application,
    isExpired,
    canSwitchToProduction:
      organization.verificationStatus === "verified" &&
      !isExpired &&
      organization.environment === "sandbox",
  };
}

export async function assertOrganizationProductionReady(organizationId: string) {
  const organization = await db.query.organizations.findFirst({
    where: eq(organizations.id, organizationId),
  });

  if (!organization) {
    throw new KycServiceError("Business not found", "not_found");
  }

  if (organization.verificationStatus !== "verified") {
    throw new KycServiceError(
      "Complete identity verification before enabling production",
      "forbidden"
    );
  }

  if (
    organization.verificationExpiresAt &&
    organization.verificationExpiresAt <= new Date()
  ) {
    throw new KycServiceError("Identity verification has expired", "forbidden");
  }

  return organization;
}

File: apps/api/internal/service/kyc/service.go
package kyc

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/payoesteam/payoes/apps/api/internal/kyc/persona"
	orgsvc "github.com/payoesteam/payoes/apps/api/internal/service/organizations"
)

const verificationValidityDays = 365

// ServiceError mirrors apps/web/src/lib/kyc/service.ts KycServiceError.
type ServiceError struct {
	Message string
	Code    string
}

func (e *ServiceError) Error() string { return e.Message }

type Service struct {
	pool   *pgxpool.Pool
	orgs   *orgsvc.Service
	persona *persona.Client
}

func NewService(pool *pgxpool.Pool, orgs *orgsvc.Service, personaClient *persona.Client) *Service {
	return &Service{pool: pool, orgs: orgs, persona: personaClient}
}

type Application struct {
	ID                   string    `json:"id"`
	OrganizationID       string    `json:"organizationId"`
	AccountType          string    `json:"accountType"`
	DisplayName          string    `json:"displayName"`
	RegistrationNumber   *string   `json:"registrationNumber"`
	Country              string    `json:"country"`
	BusinessDescription  *string   `json:"businessDescription"`
	Provider             string    `json:"provider"`
	ProviderInquiryID    *string   `json:"providerInquiryId"`
	ProviderStatus       string    `json:"providerStatus"`
	CreatedAt            time.Time `json:"createdAt"`
	UpdatedAt            time.Time `json:"updatedAt"`
}

type Summary struct {
	Organization            orgsvc.OrganizationJSON `json:"organization"`
	Application             *Application            `json:"application"`
	IsExpired               bool                    `json:"isExpired"`
	CanSwitchToProduction   bool                    `json:"canSwitchToProduction"`
}

type SessionData struct {
	InquiryID    string  `json:"inquiryId"`
	SessionToken *string `json:"sessionToken"`
}

func (s *Service) GetSummary(ctx context.Context, organizationID string) (*Summary, error) {
	org, err := s.orgs.GetByID(ctx, organizationID)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, &ServiceError{Message: "Business not found", Code: "not_found"}
	}

	app, err := s.getApplication(ctx, organizationID)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	isExpired := org.VerificationExpiresAt != nil && !org.VerificationExpiresAt.After(now)

	return &Summary{
		Organization: orgsvc.SerializeOrganization(*org),
		Application:  app,
		IsExpired:    isExpired,
		CanSwitchToProduction: org.VerificationStatus == "verified" &&
			!isExpired && org.Environment == "sandbox",
	}, nil
}

func (s *Service) StartVerification(ctx context.Context, organizationID, userID string, accountType *string) (*Application, error) {
	membership, err := s.orgs.GetMembership(ctx, organizationID, userID)
	if err != nil {
		return nil, err
	}
	if membership == nil || membership.Role != "owner" {
		return nil, &ServiceError{Message: "Only the business owner can start verification", Code: "forbidden"}
	}

	org, err := s.orgs.GetByID(ctx, organizationID)
	if err != nil {
		return nil, err
	}
	if org == nil {
		return nil, &ServiceError{Message: "Business not found", Code: "not_found"}
	}

	existing, err := s.getApplication(ctx, organizationID)
	if err != nil {
		return nil, err
	}
	if existing != nil && existing.ProviderStatus == "approved" {
		return nil, &ServiceError{Message: "This business is already verified", Code: "conflict"}
	}
	if existing != nil && existing.ProviderInquiryID != nil {
		switch existing.ProviderStatus {
		case "created", "pending", "needs_review":
			return existing, nil
		}
	}

	acctType := "personal"
	if accountType != nil && *accountType != "" {
		acctType = *accountType
	} else if existing != nil {
		acctType = existing.AccountType
	}

	note := "Personal account | Pending Persona details"
	if acctType == "business" {
		note = "Business account | Pending Persona details"
	}

	inquiry, err := s.persona.CreateInquiry(organizationID, note)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	providerStatus := persona.MapProviderStatus(inquiry.Status)

	var app *Application
	if existing != nil {
		app, err = s.scanApplication(s.pool.QueryRow(ctx, `
			UPDATE organization_verification_applications
			SET account_type = $1, display_name = $2, registration_number = NULL, country = $3,
			    business_description = NULL, provider = 'persona', provider_inquiry_id = $4,
			    provider_status = $5, updated_at = $6
			WHERE id = $7
			RETURNING id, organization_id, account_type, display_name, registration_number, country,
			          business_description, provider, provider_inquiry_id, provider_status, created_at, updated_at`,
			acctType, org.Name, "XX", inquiry.InquiryID, providerStatus, now, existing.ID))
	} else {
		app, err = s.scanApplication(s.pool.QueryRow(ctx, `
			INSERT INTO organization_verification_applications
			(organization_id, account_type, display_name, country, provider, provider_inquiry_id, provider_status, updated_at)
			VALUES ($1, $2, $3, $4, 'persona', $5, $6, $7)
			RETURNING id, organization_id, account_type, display_name, registration_number, country,
			          business_description, provider, provider_inquiry_id, provider_status, created_at, updated_at`,
			organizationID, acctType, org.Name, "XX", inquiry.InquiryID, providerStatus, now))
	}
	if err != nil {
		return nil, err
	}

	_, err = s.pool.Exec(ctx, `
		UPDATE organizations
		SET verification_status = 'pending', verified_at = NULL, verification_expires_at = NULL, updated_at = $1
		WHERE id = $2`, now, organizationID)
	if err != nil {
		return nil, err
	}
	return app, nil
}

func (s *Service) SyncFromPersona(ctx context.Context, organizationID string) (*Summary, error) {
	app, err := s.getApplication(ctx, organizationID)
	if err != nil {
		return nil, err
	}
	if app == nil || app.ProviderInquiryID == nil {
		return nil, &ServiceError{Message: "Verification has not been started yet", Code: "not_found"}
	}

	inquiry, err := s.persona.GetInquiry(*app.ProviderInquiryID)
	if err != nil {
		return nil, err
	}
	if err := s.applyInquiryProfile(ctx, app.ID, inquiry); err != nil {
		return nil, err
	}
	if err := s.applyInquiryStatus(ctx, organizationID, inquiry.InquiryID, inquiry.Status); err != nil {
		return nil, err
	}
	return s.GetSummary(ctx, organizationID)
}

func (s *Service) GetSession(ctx context.Context, organizationID, userID string) (*SessionData, error) {
	membership, err := s.orgs.GetMembership(ctx, organizationID, userID)
	if err != nil {
		return nil, err
	}
	if membership == nil {
		return nil, &ServiceError{Message: "Forbidden", Code: "forbidden"}
	}

	app, err := s.getApplication(ctx, organizationID)
	if err != nil {
		return nil, err
	}
	if app == nil || app.ProviderInquiryID == nil {
		return nil, &ServiceError{Message: "Start verification before opening the Persona flow", Code: "invalid"}
	}

	inquiry, err := s.syncInquiryState(ctx, organizationID, app)
	if err != nil {
		return nil, err
	}

	app, err = s.getApplication(ctx, organizationID)
	if err != nil || app == nil || app.ProviderInquiryID == nil {
		return nil, &ServiceError{Message: "Verification has not been started yet", Code: "not_found"}
	}

	if persona.MustBeReplaced(inquiry.Status) {
		app, err = s.replaceInquiry(ctx, organizationID, app, "Replacement inquiry after Persona declined or failed")
		if err != nil {
			return nil, err
		}
		if app.ProviderInquiryID == nil {
			return nil, &ServiceError{Message: "Verification has not been started yet", Code: "not_found"}
		}
		inquiry, err = s.persona.GetInquiry(*app.ProviderInquiryID)
		if err != nil {
			return nil, err
		}
	}

	inquiryID := *app.ProviderInquiryID
	var sessionToken *string
	if persona.NeedsSessionToken(inquiry.Status) {
		token, err := s.resumeWithFallback(ctx, organizationID, app, inquiryID, inquiry.Status)
		if err != nil {
			return nil, err
		}
		inquiryID = token.inquiryID
		sessionToken = &token.sessionToken
	}

	return &SessionData{InquiryID: inquiryID, SessionToken: sessionToken}, nil
}

func (s *Service) AssertProductionReady(ctx context.Context, organizationID string) error {
	org, err := s.orgs.GetByID(ctx, organizationID)
	if err != nil {
		return err
	}
	if org == nil {
		return &ServiceError{Message: "Business not found", Code: "not_found"}
	}
	if org.VerificationStatus != "verified" {
		return &ServiceError{Message: "Complete identity verification before enabling production", Code: "forbidden"}
	}
	if org.VerificationExpiresAt != nil && !org.VerificationExpiresAt.After(time.Now()) {
		return &ServiceError{Message: "Identity verification has expired", Code: "forbidden"}
	}
	return nil
}

type resumeResult struct {
	inquiryID    string
	sessionToken string
}

func (s *Service) resumeWithFallback(ctx context.Context, organizationID string, app *Application, inquiryID string, status persona.InquiryStatus) (resumeResult, error) {
	token, err := s.persona.ResumeInquiry(inquiryID)
	if err == nil {
		return resumeResult{inquiryID: inquiryID, sessionToken: token}, nil
	}
	if status == persona.StatusCreated || persona.MustBeReplaced(status) {
		return resumeResult{}, err
	}
	replaced, err := s.replaceInquiry(ctx, organizationID, app, "Replacement inquiry after Persona session resume failed")
	if err != nil {
		return resumeResult{}, err
	}
	if replaced.ProviderInquiryID == nil {
		return resumeResult{}, errors.New("verification has not been started yet")
	}
	token, err = s.persona.ResumeInquiry(*replaced.ProviderInquiryID)
	if err != nil {
		return resumeResult{}, err
	}
	return resumeResult{inquiryID: *replaced.ProviderInquiryID, sessionToken: token}, nil
}

func (s *Service) syncInquiryState(ctx context.Context, organizationID string, app *Application) (*persona.Inquiry, error) {
	if app.ProviderInquiryID == nil {
		return nil, &ServiceError{Message: "Verification has not been started yet", Code: "not_found"}
	}
	inquiry, err := s.persona.GetInquiry(*app.ProviderInquiryID)
	if err != nil {
		return nil, err
	}
	if err := s.applyInquiryProfile(ctx, app.ID, inquiry); err != nil {
		return nil, err
	}
	if err := s.applyInquiryStatus(ctx, organizationID, inquiry.InquiryID, inquiry.Status); err != nil {
		return nil, err
	}
	return inquiry, nil
}

func (s *Service) replaceInquiry(ctx context.Context, organizationID string, app *Application, note string) (*Application, error) {
	if note == "" {
		if app.AccountType == "business" {
			note = "Business account | New Persona inquiry"
		} else {
			note = "Personal account | New Persona inquiry"
		}
	}
	inquiry, err := s.persona.CreateInquiry(organizationID, note)
	if err != nil {
		return nil, err
	}
	now := time.Now()
	replaced, err := s.scanApplication(s.pool.QueryRow(ctx, `
		UPDATE organization_verification_applications
		SET provider_inquiry_id = $1, provider_status = $2, updated_at = $3
		WHERE id = $4
		RETURNING id, organization_id, account_type, display_name, registration_number, country,
		          business_description, provider, provider_inquiry_id, provider_status, created_at, updated_at`,
		inquiry.InquiryID, persona.MapProviderStatus(inquiry.Status), now, app.ID))
	if err != nil {
		return nil, err
	}
	_, err = s.pool.Exec(ctx, `
		UPDATE organizations
		SET verification_status = 'pending', verified_at = NULL, verification_expires_at = NULL, updated_at = $1
		WHERE id = $2`, now, organizationID)
	return replaced, err
}

func (s *Service) applyInquiryProfile(ctx context.Context, applicationID string, inquiry *persona.Inquiry) error {
	displayName, country := persona.ExtractProfile(inquiry.Fields)
	if displayName == nil && country == nil {
		return nil
	}
	if country != nil {
		normalized := normalizeCountryCode(*country)
		country = &normalized
	}
	_, err := s.pool.Exec(ctx, `
		UPDATE organization_verification_applications
		SET display_name = COALESCE($1, display_name),
		    country = COALESCE($2, country),
		    updated_at = NOW()
		WHERE id = $3`, displayName, country, applicationID)
	return err
}

func (s *Service) applyInquiryStatus(ctx context.Context, organizationID, inquiryID string, status persona.InquiryStatus) error {
	app, err := s.getApplication(ctx, organizationID)
	if err != nil || app == nil || app.ProviderInquiryID == nil || *app.ProviderInquiryID != inquiryID {
		return err
	}

	providerStatus := persona.MapProviderStatus(status)
	verificationStatus := persona.MapVerificationStatus(status)
	now := time.Now()

	_, err = s.pool.Exec(ctx, `
		UPDATE organization_verification_applications
		SET provider_status = $1, updated_at = $2
		WHERE id = $3`, providerStatus, now, app.ID)
	if err != nil {
		return err
	}

	switch verificationStatus {
	case "verified":
		expires := now.AddDate(0, 0, verificationValidityDays)
		_, err = s.pool.Exec(ctx, `
			UPDATE organizations
			SET verification_status = 'verified', verified_at = $1, verification_expires_at = $2, updated_at = $1
			WHERE id = $3`, now, expires, organizationID)
	case "rejected":
		_, err = s.pool.Exec(ctx, `
			UPDATE organizations
			SET verification_status = 'rejected', verified_at = NULL, verification_expires_at = NULL, updated_at = $1
			WHERE id = $2`, now, organizationID)
	default:
		_, err = s.pool.Exec(ctx, `
			UPDATE organizations SET verification_status = 'pending', updated_at = $1 WHERE id = $2`, now, organizationID)
	}
	return err
}

func (s *Service) getApplication(ctx context.Context, organizationID string) (*Application, error) {
	return s.scanApplication(s.pool.QueryRow(ctx, `
		SELECT id, organization_id, account_type, display_name, registration_number, country,
		       business_description, provider, provider_inquiry_id, provider_status, created_at, updated_at
		FROM organization_verification_applications
		WHERE organization_id = $1 LIMIT 1`, organizationID))
}

func (s *Service) scanApplication(row pgx.Row) (*Application, error) {
	var a Application
	err := row.Scan(
		&a.ID, &a.OrganizationID, &a.AccountType, &a.DisplayName, &a.RegistrationNumber, &a.Country,
		&a.BusinessDescription, &a.Provider, &a.ProviderInquiryID, &a.ProviderStatus, &a.CreatedAt, &a.UpdatedAt,
	)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &a, nil
}

func normalizeCountryCode(country string) string {
	trimmed := strings.TrimSpace(country)
	if len(trimmed) == 2 {
		return strings.ToUpper(trimmed)
	}
	if len(trimmed) >= 2 {
		return strings.ToUpper(trimmed[:2])
	}
	return strings.ToUpper(trimmed)
}
